### PR TITLE
fix(mailroom): reject post requests over capacity

### LINKED ISSUE #1507: mailroom: check capacity when write to `Mailboxes`
### Which crate is this feature request for (if any)?

payjoin-mailroom

### Question

I was looking at `payjoin-mailroom/src/db/files.rs` and noticed that `has_capacity()` is checked in `wait_v2` before inserting into `pending_v2`, but `post_v2` does not check before inserting into `insert_order`. 

Both contribute to [`len()`](https://github.com/payjoin/rust-payjoin/blob/fc665ebbb48bc981259cb4e64222a7301a49bf02/payjoin-mailroom/src/db/files.rs#L431-L433):
```rust
fn len(&self) -> usize {
    self.insert_order.len() + self.pending_v1.len() + self.pending_v2.len()
}
```

[`wait_v2`](https://github.com/payjoin/rust-payjoin/blob/fc665ebbb48bc981259cb4e64222a7301a49bf02/payjoin-mailroom/src/db/files.rs#L333-L339) checks before allocating a waiter:
```rust
async fn wait_v2(...) -> ... {
    if !self.has_capacity().await? {
        return Err(Error::OverCapacity);
    }
    // ... inserts into self.pending_v2
}
```

[`post_v2`](https://github.com/payjoin/rust-payjoin/blob/fc665ebbb48bc981259cb4e64222a7301a49bf02/payjoin-mailroom/src/db/files.rs#L370-L375) writes to disk and pushes to `insert_order` without a capacity check:
```rust
async fn post_v2(&mut self, id: &ShortId, payload: Vec<u8>) -> Result<Option<()>, Error> {
    let Some(created) = self.persistent_storage.try_insert(id, &payload).await? else {
        return Ok(None);
    };
    self.insert_order.push_back((created, *id));
    // ...
}
```

[`DEFAULT_CAPACITY`](https://github.com/payjoin/rust-payjoin/blob/fc665ebbb48bc981259cb4e64222a7301a49bf02/payjoin-mailroom/src/db/files.rs#L18-L22) is meant as a global cap on pending + populated entries:
```rust
/// The maximum number of pending or populated mailbox entries.
///
/// Defaults to around 2e6, for a generous upper bound rounded up from ~2
/// mailboxes/tx, ~4K txs/block, and ~144 blocks/24h.
const DEFAULT_CAPACITY: usize = 1 << (1 + 12 + 8);
```

Practical consequence: a flood of `post_v2` calls can grow `insert_order` (and disk usage) past `DEFAULT_CAPACITY`, while legitimate `wait_v2` callers start getting `OverCapacity`.

Happy to send a PR adding `has_capacity()` to `post_v2` - unless there's a reason I'm missing for the asymmetry.

### PR DESCRIPTION
**Pull request by va-an**
Closes #1507 

Both `post_v2` and `post_v1_req_and_wait` grow the in-memory counters tracked by `len()`, but only `wait_v2` was guarded by `has_capacity()`. This patch applies the same check to the two write paths so the capacity bound is enforced symmetrically.

<details>
  <summary>Pull Request Checklist</summary>

Please confirm the following before requesting review:

- [x] I have [disclosed my use of AI](https://github.com/payjoin/rust-payjoin/blob/master/.github/CONTRIBUTING.md#ai-assistance-notice) in the body of this PR.
- [x] I have read [CONTRIBUTING.md](https://github.com/payjoin/rust-payjoin/blob/master/.github/CONTRIBUTING.md#commits) and **rebased my branch to produce [hygienic commits](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#committing-patches)**.
</details>

### CODE REVIEW

**Review by nothingmuch — APPROVED** [review]
utACK, not sure how I missed this (when i wrote it)

i would be happier if this included regression tests similar to the existing capacity tests

File: payjoin-mailroom/src/db/files.rs [hunk 1]
```diff
@@ -392,6 +396,10 @@ impl Mailboxes {
         id: &ShortId,
         payload: Vec<u8>,
     ) -> Result<Option<oneshot::Receiver<Vec<u8>>>, Error> {
+        if !self.has_capacity().await? {
```
Comment (nothingmuch) [inline payjoin-mailroom/src/db/files.rs:401]: note that technically in this fn it could be that the v2 removal makes room for the v1 insert.

i think it's better this way since the code is simpler, but technically it implies v1 writers will in some cases see effective capacity of n-1 instead of n

Comment (vadim-anfv) [reply payjoin-mailroom/src/db/files.rs:401]: That's not obvious, thanks for pointing it out. 
I'll come back to this when I understand the code better.

**Comment by vadim-anfv** [comment]
> i would be happier if this included regression tests similar to the existing capacity tests

Thanks for the review. 
Couldn't find existing capacity tests to mirror, so I added one per guarded DB method.

File: payjoin-mailroom/src/db/files.rs [hunk 2]
```diff
@@ -939,3 +947,95 @@ async fn test_prune() -> std::io::Result<()> {
 
     Ok(())
 }
+
+fn random_id() -> ShortId { ShortId(OsRng.next_u64().to_ne_bytes()) }
+
+fn random_payload() -> Vec<u8> { rand::random::<[u8; 32]>().to_vec() }
+
```
Comment (caarloshenriq) [inline payjoin-mailroom/src/db/files.rs:941]: When a ran using `cargo run` The compiler returns two warnings about these functions being unused. Since they are only used in tests, you can use `#[cfg(test)]`
```suggestion

#[cfg(test)]
fn random_id() -> ShortId { ShortId(OsRng.next_u64().to_ne_bytes()) }

#[cfg(test)]
fn random_payload() -> Vec<u8> { rand::random::<[u8; 32]>().to_vec() }

```

File: payjoin-mailroom/src/db/files.rs (hunk 1 above)
Comment (nothingmuch) [reply payjoin-mailroom/src/db/files.rs:401]: IMO don't bother fixing this, maybe just add a comment to indicate that this edge case exists

**Comment by nothingmuch** [comment]
> Thanks for the review. Couldn't find existing capacity tests to mirror, so I added one per guarded DB method.

`test_prune` is the test i had in mind FWIW

File: payjoin-mailroom/src/db/files.rs (hunk 2 above)
Comment (nothingmuch) [reply payjoin-mailroom/src/db/files.rs:941]: in the rest of the codebase when we have this kind of stuff we usually add

```rs
#[cfg(test)]
mod tests {
    ...
}
```

around all the tests to reduce the boilerplate, i can't decide if it's worth it in this case

File: payjoin-mailroom/src/db/files.rs [hunk 3]
```diff
@@ -939,3 +947,95 @@ async fn test_prune() -> std::io::Result<()> {
 
     Ok(())
 }
+
+fn random_id() -> ShortId { ShortId(OsRng.next_u64().to_ne_bytes()) }
+
+fn random_payload() -> Vec<u8> { rand::random::<[u8; 32]>().to_vec() }
```
Comment (nothingmuch) [inline payjoin-mailroom/src/db/files.rs]: i think these functions can be removed, you can just use the loop index as the id

File: payjoin-mailroom/src/db/files.rs [hunk 4]
```diff
@@ -939,3 +947,95 @@ async fn test_prune() -> std::io::Result<()> {
...
+    db.mailboxes.lock().await.capacity = 3;
+
+    for _ in 1..=3 {
+        db.post_v2_payload(&random_id(), random_payload()).await.unwrap();
+    }
+
+    assert!(matches!(
+        db.post_v2_payload(&random_id(), random_payload()).await,
+        Err(DbError::OverCapacity)
+    ));
```
Comment (nothingmuch) [inline payjoin-mailroom/src/db/files.rs]: all mailboxes share the same capacity, so i think it'd be more correct to assert that whenever there's no capacity, this is true of all request types:

```suggestion
    assert!(matches!(
        db.post_v2_payload(&random_id(), random_payload()).await,
        Err(DbError::OverCapacity)
    ));
    
    assert!(matches!(db.wait_for_v2_payload(&random_id()).await, Err(DbError::OverCapacity)));
    
    assert!(matches!(
        db.post_v1_request_and_wait_for_response(&random_id(), random_payload()).await,
        Err(DbError::OverCapacity)
    ));
```

same for the lower checks (so this can be a utility function, or even just a single test which always asserts this for a variety of different at-capacity setups)

File: payjoin-mailroom/src/db/files.rs (hunk 2 above)
Comment (caarloshenriq) [reply payjoin-mailroom/src/db/files.rs:941]: Agreed, wrapping everything in a #[cfg(test)] mod tests { } block is cleaner and more consistent with the rest of the codebase. It avoids annotating each helper individually and makes it clear at a glance that all of it is test-only code

Comment (vadim-anfv) [reply payjoin-mailroom/src/db/files.rs:941]: I would prefer to wrap tests in `files.rs` to `mod tests {}` for consistency with codebase, even we decided to delete this helper functions below.

Comment (vadim-anfv) [reply payjoin-mailroom/src/db/files.rs:941]: @nothingmuch should I add commit hash d8afeaabe75572713dae024d9b2676a05ec1053c to `.git-blame-ignore-revs` file? And if so - what's the accepted way to do this: a commit in this PR, a separate PR, etc?

Comment (nothingmuch) [reply payjoin-mailroom/src/db/files.rs:941]: > should I add commit hash [d8afeaa](https://github.com/payjoin/rust-payjoin/commit/d8afeaabe75572713dae024d9b2676a05ec1053c) to `.git-blame-ignore-revs` file?

no need IMO, git blame has -w option, ignore-revs is more useful for massive reformatting that isn't trivially addressed by git's move/copy detection or -w

>  And if so - what's the accepted way to do this: a commit in this PR, a separate PR, etc?

in general, just another commit in the PR makes the most sense to me, it's impossible to do it in one commit but that's the closest thing to that

**Review by nothingmuch — APPROVED** [review]
utACK

**Review by caarloshenriq — APPROVED** [review]
tACK 48cc7c7

**Review by spacebear21 — APPROVED** [review]
ACK

**Comment by spacebear21** [comment]
@va-an sorry for the delay in running the CI workflows, can you fix the lint errors before we merge?

**Comment by vadim-anfv** [comment]
> @va-an sorry for the delay in running the CI workflows, can you fix the lint errors before we merge?

Done - 24840c5812202c095ac8bbf41b9a429a32815149. These lint errors didn't show up locally under devenv.
Want me to squash into a single commit before merging?

**Comment by nothingmuch** [comment]
> These lint errors didn't show up locally under devenv. Want me to squash into a single commit before merging?

i think this is the result of the toolchain (provided by `dtolnay/rust-toolchain@nightly` action) bumping clippy version, it's not related to your PR at all

**Review by spacebear21 — APPROVED** [review]
reACK
